fuse/src/main/java/tachyon/fuse/TachyonFuseFs.java

```diff
@@ -0,0 +1,577 @@
+
+        // Assuming I will never wrap around (2^64 open files are quite a lot anyway)
+        mOpenFileIds += 1;
+      }
+      LOG.debug("{} created and opened in O_WRONLY mode", path);
+
+    } catch (FileAlreadyExistsException e) {
+      LOG.debug("File {} already exists", turi, e);
+      ret = ErrorCodes.EEXIST();
+    } catch (IOException e) {
+      LOG.error("IOException on {}", path, e);
+      ret = ErrorCodes.EFAULT();
+    } catch (TachyonException e) {
+      LOG.error("TachyonException on {}", path, e);
+      ret = ErrorCodes.EFAULT();
+    } catch (Throwable e) {
+      LOG.error("Unexpected exception on {}", path, e);
+      ret = ErrorCodes.EFAULT();
+    }
+
+    return -ret;
+  }
+
+  @Override
+  public int flush(String path, FuseFileInfo fi) {
+    LOG.trace("flush({})", path);
+    final long fd = fi.fh.get();
+    OpenFileEntry oe = null;
+    synchronized (mOpenFilesLock) {
+      oe = mOpenFiles.get(fd);
+    }
+    if (oe == null) {
+      LOG.error("Cannot find fd for {} in table", path);
+      return -ErrorCodes.EBADFD();
+    }
+    if (oe.mOut == null) {
+      LOG.error("{} was not open for writing", path);
+      return -ErrorCodes.EBADFD();
+    }
+    try {
+      oe.mOut.flush();
+    } catch (IOException e) {
+      return -ErrorCodes.EIO();
+    }
+
+    return 0;
+  }
+
+  @Override
+  public int getattr(String path, FileStat stat) {
+    int ret = 0;
+    final TachyonURI turi = mPathResolverCache.getUnchecked(path);
+    LOG.trace("getattr({}) [Tachyon: {}]", path, turi);
+    try {
+      final TachyonFile tf = mTFS.openIfExists(turi);
+      if (tf == null) {
+        ret = ErrorCodes.ENOENT();
+      } else {
+        final FileInfo fi = mTFS.getInfo(tf);
+        stat.st_size.set(fi.getLength());
+
+        final long ctime = fi.getLastModificationTimeMs();
+        final long ctime_sec = fi.getLastModificationTimeMs() / 1000;
+        final long ctime_nsec = (fi.getLastModificationTimeMs() % 1000) * 1000;
+        stat.st_ctim.tv_sec.set(ctime_sec);
+        stat.st_ctim.tv_nsec.set(ctime_nsec);
+        stat.st_mtim.tv_sec.set(ctime_sec);
+        stat.st_mtim.tv_nsec.set(ctime_nsec);
+
+        System.getProperty("user.name");
+
+        stat.st_uid.set(UID_AND_GID[0]);
+        stat.st_gid.set(UID_AND_GID[1]);
+
+        final int mode;
+        if (fi.isFolder) {
+          mode = FileStat.S_IFDIR;
+        } else {
+          mode = FileStat.S_IFREG;
+        }
+        stat.st_mode.set(mode);
+      }
+
+    } catch (InvalidPathException e) {
+      LOG.debug("Invalid path {}", path, e);
+      ret = ErrorCodes.ENOENT();
+    } catch (FileDoesNotExistException e) {
+      LOG.debug("File does not exist {}", path, e);
+      ret = ErrorCodes.ENOENT();
+    } catch (IOException e) {
+      LOG.error("IOException on {}", path, e);
+      ret = ErrorCodes.EFAULT();
+    } catch (TachyonException e) {
+      LOG.error("TachyonException on {}", path, e);
+      ret = ErrorCodes.EFAULT();
```

Comment:
Perhaps we can give different error codes based on the specific type of TachyonException.